### PR TITLE
Add pushdown for parquet timestamp predicate

### PR DESCRIPTION
**Pull request by alexjo2144**

### CODE REVIEW

File: presto-parquet/src/main/java/io/prestosql/parquet/predicate/ParquetTimestampStatistics.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+public class ParquetTimestampStatistics
+        implements ParquetRangeStatistics<Long>
+{
+    private final Long minimum;
+    private final Long maximum;
+
+    public ParquetTimestampStatistics(Long min, Long max)
+    {
+        this.minimum = min;
+        this.maximum = max;
```
Comment (findepi) [inline presto-parquet/src/main/java/io/prestosql/parquet/predicate/ParquetTimestampStatistics.java:25]: You can make fields and ctor args primitive `long`

File: presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java [hunk 2]
```diff
@@ -315,6 +322,55 @@ public void testDate()
                 .withMessage("Corrupted statistics for column \"DateColumn\" in Parquet file \"testFile\": [min: 200, max: 100, num_nulls: 0]");
     }
 
+    @Test
+    public void testTimestampMilli()
```
Comment (findepi) [inline presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java:326]: testTimestampMilli -> testTimestampMillis

File: presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java [hunk 3]
```diff
@@ -315,6 +322,55 @@ public void testDate()
...
+            throws ParquetCorruptionException
+    {
+        Instant now = Instant.now();
```
Comment (findepi) [inline presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java:329]: use constant moment

File: presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java [hunk 4]
```diff
@@ -315,6 +322,55 @@ public void testDate()
...
+    {
+        long startOfDay = LocalDate.ofInstant(timestamp, ZoneOffset.UTC).atStartOfDay().toInstant(ZoneOffset.UTC).toEpochMilli();
+        long timeOfDayNanos = (long) ((timestamp.toEpochMilli() - startOfDay) * Math.pow(10, 6));
+
+        ByteBuffer encoding = ByteBuffer.allocate(12).order(ByteOrder.LITTLE_ENDIAN);
+        encoding.putLong(timeOfDayNanos);
+        encoding.putInt(millisToJulianDay(timestamp.toEpochMilli()));
+        byte[] arr = new byte[12];
+        encoding.position(0).get(arr, 0, 12);
+        return arr;
```
Comment (findepi) [inline presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java:363]:
```
Slice slice = Slices.allocate(12);
slice.setLong(0, timeOfDayNanos);
slice.setInt(8, millisToJulianDay(timestamp.toEpochMilli()));
return slice.byteArray();
```

File: presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java [hunk 5]
```diff
@@ -315,6 +322,55 @@ public void testDate()
...
+    }
+
+    private static final int JULIAN_EPOCH_OFFSET_DAYS = 2_440_588;
+    private static final long MILLIS_IN_DAY = TimeUnit.DAYS.toMillis(1);
+
+    private static int millisToJulianDay(long timestamp)
+    {
+        return (int) (timestamp / MILLIS_IN_DAY) + JULIAN_EPOCH_OFFSET_DAYS;
```
Comment (findepi) [inline presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java:371]: - we usually just `[TimeUnit.]MILLISECONDS.toDays(timestamp)`
- avoid unckeched `(int)` cast

```suggestion
        return toIntExact(MILLISECONDS.toDays(timestamp) + JULIAN_EPOCH_OFFSET_DAYS);
```

File: presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java [hunk 6]
```diff
@@ -315,6 +322,55 @@ public void testDate()
...
+
+        ByteBuffer encoding = ByteBuffer.allocate(12).order(ByteOrder.LITTLE_ENDIAN);
+        encoding.putLong(timeOfDayNanos);
+        encoding.putInt(millisToJulianDay(timestamp.toEpochMilli()));
+        byte[] arr = new byte[12];
+        encoding.position(0).get(arr, 0, 12);
+        return arr;
+    }
+
+    private static final int JULIAN_EPOCH_OFFSET_DAYS = 2_440_588;
```
Comment (findepi) [inline presto-parquet/src/test/java/io/prestosql/parquet/TestTupleDomainParquetPredicate.java:366]: Let's reuse `io.prestosql.parquet.ParquetTimestampUtils#JULIAN_EPOCH_OFFSET_DAYS`
(package private, `@VisibleForTesting`)

**Comment by alexjo2144** [comment]
Updated, thanks

**Comment by findepi** [comment]
Merged, thanks!

**Comment by ryanrupp** [comment]
Does this cover `int64` logically typed as `TIMESTAMP_MICROS` OR `TIMESTAMP_MILLIS` (also timestamp nanos I believe in newer versions of Parquet)? I wasn't familiar with `ParquetTimestampUtils` which is used here but it seems to reference `int96` timestamps which is deprecated for Parquet. Also this is using `BinaryStatistics` then (probably what you get for int96 encoded timestamps? not familiar).

For reference there was an older PR #1999 that was adding pushdown for timestamps on `int64` via `LongStatistics`

**Comment by alexjo2144** [comment]
Hey Ryan, thanks for pointing that out. I think you're right, this only works for the legacy int96 timestamps. I didn't realize it was deprecated.

Do you know if Presto's Parquet read/writer supports the new format or should we add an issue to add that in? There's probably some more changed needed to work in the parametric timestamp types that Dain has been working on too.

**Comment by ryanrupp** [comment]
@alexjo2144 I'm not sure what the writer actually writes out currently (my write path isn't through Presto currently). For the read path I'm pretty sure it can read int64 TIMESTAMP_MILLIS but I'm not sure about int64 logically typed as TIMESTAMP_MICROS. I thought of this because there was a post in Slack asking if TIMESTAMP_MICROS is supported on the read path [here](https://prestosql.slack.com/archives/CP1MUNEUX/p1594620479158000)

[Here's](https://github.com/apache/parquet-format/blob/345282ce307a9c9dcc15b6f3c9106be379ec26ba/LogicalTypes.md#timestamp) the documentation about timestamp handling in Parquet. One of the newer features is support for nanos precision stored in an int64 (with nano precision in int96 being deprecated, see [here](https://issues.apache.org/jira/browse/PARQUET-323))

**Comment by alexjo2144** [comment]
Gotcha. Looks like the default presto parquet writer uses int96, but there's a new writer, added the same day as these changes, that supports int64 at millisecond precision https://github.com/prestosql/presto/pull/3400 That should help for testing.

I can work on putting up an updated version of https://github.com/prestosql/presto/pull/1999 to tie that in.

**Comment by ericxiao251** [comment]
This improvement really helped with one of our queries, but I want to understand a bit more of what's actually happening with Presto and taking advantage of the parquet metadata...

What does the `getDomain` function in the `TupleDomainParquetPredicate.java` file do? How does leveraging timestamp metadata help when data is randomly stored on disk?

**Comment by alexjo2144** [comment]
So, Parquet files are separated out into `RowGroups`. Each RowGroup gets an entry in the file footer with statistics information. This section of the code reads that footer and determines which sections of the file can be skipped by comparing the footer statistics with any filters from the query. `getDomain` returns that statistics information as a `Domain` that the engine can use for this comparison. Skipping whole groups means being able to reduce IO, and also do less filtering after the data is loaded into the engine.

**Pull request merged**